triage/summarize/files.go

```diff
@@ -0,0 +1,396 @@
+		}
+		b.Elapsed = int(tempElapsed)
+	}
+
+	// testsRun
+	if jb.TestsRun != "" {
+		b.TestsRun, err = strconv.Atoi(jb.TestsRun)
+		if err != nil {
+			return build{}, fmt.Errorf("Error converting JSON string '%s' to int for build field 'testsRun': %s", jb.TestsRun, err)
+		}
+	}
+
+	// testsFailed
+	if jb.TestsFailed != "" {
+		b.TestsFailed, err = strconv.Atoi(jb.TestsFailed)
+		if err != nil {
+			return build{}, fmt.Errorf("Error converting JSON string '%s' to int for build field 'testsFailed': %s", jb.TestsFailed, err)
+		}
+	}
+
+	// number
+	if jb.Number != "" {
+		b.Number, err = strconv.Atoi(jb.Number)
+		if err != nil {
+			return build{}, fmt.Errorf("Error converting JSON string '%s' to int for build field 'number': %s", jb.Number, err)
+		}
+	}
+
+	return b, nil
+}
+
+// loadBuilds parses a JSON file containing build information and returns a map from build paths
+// to build objects.
+func loadBuilds(filepath string) (map[string]build, error) {
+	// The map
+	builds := make(map[string]build)
+
+	// jsonBuilds temporarily stores the builds as they are retrieved from the JSON file
+	// until they can be converted to build objects
+	jsonBuilds := make([]jsonBuild, 0)
+
+	err := getJSON(filepath, &jsonBuilds)
+	if err != nil {
+		return nil, fmt.Errorf("Could not get builds JSON: %s", err)
+	}
+
+	// Convert the build information to internal build objects and store them in the builds map
+	for _, jBuild := range jsonBuilds {
+		// Skip builds without a start time or build number
+		if jBuild.Started == "" || jBuild.Number == "" {
+			continue
+		}
+
+		bld, err := jBuild.asBuild()
+		if err != nil {
+			return nil, fmt.Errorf("Could not create build object from jsonBuild object: %s", err)
+		}
+
+		if strings.Contains(bld.Path, "pr-logs") {
+			parts := strings.Split(bld.Path, "/")
+			bld.PR = parts[len(parts)-3]
+		}
+
+		builds[bld.Path] = bld
+	}
+
+	return builds, nil
+}
+
+// jsonFailure represents a test failure as reported by the JSON. All values are strings.
+// This should not be instantiated directly, but rather via the encoding/json package's
+// Unmarshal method. This is an intermediary state for the data until it can be put into
+// a failure object.
+type jsonFailure struct {
+	Started     string `json:"started"`
+	Build       string `json:"build"`
+	Name        string `json:"name"`
+	FailureText string `json:"failure_text"`
+}
+
+// asFailure is a factory function that creates a failure object from the jsonFailure object,
+// appropriately handling all type conversions.
+func (jf *jsonFailure) asFailure() (failure, error) {
+	// The failure object that will be returned, initialized with the values that
+	// don't need conversion.
+	f := failure{
+		Build:       jf.Build,
+		Name:        jf.Name,
+		FailureText: jf.FailureText,
+	}
+
+	// To avoid assignment issues
+	var err error
+
+	// started
+	if jf.Started != "" {
```

Comment:
As in, make the function call and then check for the empty string? I think that would take longer

Reply:
Just mean before creating the base failure. But really doesn't make a difference. I just like the failure case as high in the method as I can get it

Reply:
I understand. I think in this case it would make it more complicated as it would necessitate a temp variable to hold the result while checking the error, and then copying over that temp value in f.Started